### ISSUE TITLE
 study apt-p2p study for updates

### ISSUE THREAD

**Issue by orwellencarnation** [Librerouter/Librekernel#178]

**Comment by luisnarvaez19** [comment]
The Apt-P2P daemon runs automatically on startup and listens for requests from APT for files, as well as maintaining membership in a Distributed Hash
 Table (DHT). For any files which have a hash available (most files), peers that have the file are looked for in the DHT. The file can then be downloaded 
 from them, using the uploading bandwidth of other peers while reducing the demand on the Debian mirror network. However, if a package can not be 
 found on any peers, Apt-P2P will fall back to downloading from a mirror to ensure all packages are downloaded. Once downloaded, the file is hash 
 checked and added to the DHT so that other peers can download it without the mirror.
The configuration is very simple, and only involves prepending 'localhost' and the port to your current sources.list lines (similar to other apt-proxy-like software).

Features (Pros)

1. Greater privacy since main repositories are not centralized
2. Downloads from peers, increasing the available bandwidth to the user
3. Reduces the bandwidth requirements needed to setup a repository of packages
4. Seamlessly integrates with the current APT tool
5. Automatically falls back to downloading from an HTTP mirror when peers are not available
6. Builds on other already existing tools where possible
7. Fast and requires limited CPU and memory
8. Will try to download any file it can find a hash for from peers (including Packages.bz2, Sources.gz, ...)

**Comment by luisnarvaez19** [comment]
Cons
 
 1. It works but network traffic doesn't stop with it.
 2. As every P2p network increases the use of ram in the computer
 3. When you kill the process and make a top before and after you can notice the change. So much in the use of the CPU and much more 
 in the one of the RAM memory
 4.  About the network traffic: wow, in the meantime my bandwidth got from 1024/128 Kbps to 4096/256 Kbps and I can finally see the benefits since
  those torrents used a big part of my upload rate and ... never mind now its solved 
 5.  Are the owners of the main repositories interested in my data?
 6.  Verifying an update as authentic could be easily done and a minor concern. I would be more concerned to having a system open to 24/7 p2p 
 access for something as critical as OS updates. You know someone will find a way to push in unwanted updates/code.
 7. The protocol is designed to distribute large amounts of data. Updates are not considered large amounts of data as each update is typically
  a few hundred Kb. Then there would be the creation, indexing and tracking of each package. It is a lot of extra stuff to maintain for little added 
  benefit (decentralized updates) that are already taken care of by having multiple mirrors.
  8. There's also an issue with security. Anyone can join a public BitTorrent swarm and get a list of peers who are requesting a specific file. If this 
  file is a critical security update, it's possible for an attacker to obtain multiple targets from the swarm and pull off attack(s) on them before
   they've applied the update or before the vulnerable components have been reloaded (in cases where a program restart or even a full reboot 
   is necessary).
  9. The entire network is also not secure. This is because peers can transmit spywares, viruses, malwares and Trojans. These threats can spread 
  from one peer to the other very easily since in most cases nobody monitors the system. 
  
   
 Suggestions
 
 1. If you decide to install it you should use GPG checking, to ensure that the repositories are authentic

**Comment by luisnarvaez19** [comment]
We would like to know your opinion about whether we should implement apt-p2p in Librerouter / Librekernel. Please read the pros and cons in issue 178. We appreciate your comments.

**Comment by luisnarvaez19** [comment]
Comments from Franco:

Hi Luis,


Cons
1 . About traffic bandwidth utilization, same here, as the project I am working on is over Tahoe+Tor
     and requires nodes still ready on some portion
     That means even my node is not doing restoring/backing, I am sharing space that must be ready
     for other Tahoe nodes on other librerouter boxes when they do backup/restore

3. Same on Tahoe, when process is started takes a lot of CPU for a while.
    I have need to check it before to allow backup/restore tasks

4. On tahoe grid, bw is good controlled

9. That risk will be always , is right can be increased if p2p services, but I would not come concerned
    about.

May be not bad idea to implementn p2p over tahoe too. With a little hacking I success on most of
tahoe tasks.
I have left to do some metadata2file to save/restore files by chunks CRC as well as to conserve the
original metadata ( related with your point 7 ) , in a way stored files must have some extra information,
as is impossible to do CRC on each block on both sides and I did over sshfs in the same side, that is
not very good, you save upload bandwidth at cost to convert it on download data ( that in most cases
is a little better )

About privacity, I success running tahoe introducer over Tor, as well as all nodes over Tor with enough
good speed.  I have not success yet over i2p ( pending some tasks from tahoe developers guys )

Also Tahoe may be a good solution, I did my project with two nodes, one public ( shared between all
librerouter ) and one private ( where really the backups are stored ) , but may be including other new
node on same box can help on p2p based services.

My tasks on this backup/restore project are on the repository if you like to give a sight ( for now all
is pure bash )

Hope this helps

Joaquin F.
34 722330948
